<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Promise</title>
</head>

<body>
  <script>
    const PENDING_STATE = 'pending';
    const FULFILLED_STATE = 'fulfilled';
    const REJECTED_STATE = 'rejected';

    class Promise {

      constructor(executor) {
        this.state = PENDING_STATE; // 初始化状态
        this.result = undefined;
        this.onResolvedCallbacks = []; // 存放当成功时，需要执行的函数列表
        this.onRejectedCallbacks = []; // 存放当失败时，需要执行的函数列表

        const resolve = (value) => {
          if (this.state === PENDING_STATE) {
            this.state = FULFILLED_STATE;
            this.result = value;

            this.onResolvedCallbacks.forEach(fn => fn())
          }
        };

        const reject = (reason) => {
          if (this.state === PENDING_STATE) {
            this.state = REJECTED_STATE;
            this.result = reason;

            this.onRejectedCallbacks.forEach(fn => fn())
          }
        };

        // 执行executor报错，直接reject
        try {
          executor(resolve, reject);
        } catch (error) {
          reject(error);
        }
      }

      then(onFulFilled, onRejected) {
        const promise = new Promise((resolve, reject) => {
          // onFulFilled 如果成功了，应该被调用
          // onRejected  如果失败了，应该被调用
          if (this.state === FULFILLED_STATE) {
            setTimeout(() => {
              try {
                const result = onFulFilled(this.result)
                // 分析result
                // 如果是promise对象 看promise是否成功 如果成功就resolve 如果失败就reject
                // 如果是普通值，直接返回
                resolvePromise(result, resolve, reject, promise)
              } catch (error) {
                reject(error)
              }
            }, 0);
          }

          if (this.state === REJECTED_STATE) {
            setTimeout(() => {
              try {
                const result = onRejected(this.result)
                rejectPromise(result, resolve, reject, promise)
              } catch (error) {
                reject(error)
              }
            }, 0);
          }

          // 处理异步的情况
          if (this.state === PENDING_STATE) {
            this.onResolvedCallbacks.push(() => {
              onFulFilled(this.result)
            })

            this.onRejectedCallbacks.push(() => {
              onRejected(this.result)
            })
          }
        })
        return promise
      }

      static resolve(val) {

        return new Promise((resolve, reject) => {
          resolve(val)
        })
      }

      static reject(val) {

        return new Promise((resolve, reject) => {
          reject(val)
        })
      }

      // 等待原则， 传入多个promise，等所有的promise都满足条件，拿到所有的成功结果
      static all(promises) {
        const arr = []
        let i = 0 // 必须要用计数器判断，累计有多少次成功了， 如果结果的个数和promises个数相等，满足条件

        return new Promise((resolve, reject) => {
          promises.forEach((promise, index) => {
            promise.then(res => {
              arr[index] = res
              i++
              if (promises.length === i) { // 不能用promises.length === arr.length 原因是刚开始遇到异步的arr的数组长度等于promises.length，但是里面的异步的值为空
                resolve(arr)
              }
            }, reject)
          })
        })
      }

      // 竞速原则，谁选满足条件，就先被.then处理，其它的就忽略，返回最快的结果
      static race(promises) {

        return new Promise((resolve, reject) => {
          promises.forEach(promise => {
            promise.then(resolve, reject)
          })
        })
      }
    }

    const resolvePromise = (result, resolve, reject, promise) => {
      if (result === promise) {
        throw new TypeError('Chaining cycle detected for promise #<Promise>')
      }

      if (result instanceof Promise) {
        result.then(res => resolve(res), error => reject(error))
      } else {
        resolve(result)
      }
    }

    const rejectPromise = (result, resolve, reject, promise) => {
      if (result === promise) {
        throw new TypeError('Chaining cycle detected for promise #<Promise>')
      }

      if (result instanceof Promise) {
        result.then(res => resolve(res), error => reject(error))
      } else {
        reject(result)
      }
    }

    // promiseA+ (resolve, reject) => {} 是executor
    const p = new Promise((resolve, reject) => {
      // resolve 和 reject 是 Promise 内部实现好的函数
      // 这里的代码是立即执行的

      // reject('失败'); // 将状态从 pending 改成了 rejected （状态凝固）
      // setTimeout(() => {
      //   resolve(200); // 将状态从 pending 改成了 fulfilled
      // }, 1000);
      reject(500)
    })

    // onFulFilled => p成功后，调用的回调
    // onRejected =>  p失败后，调用的回调
    // .catch可以全局捕获错误
    // p.then(onFulFilled, onRejected).catch()
    // p.then(res => {
    //   console.log('res: ', res);
    //   // return '成功'
    //   return new Promise((resolve, reject) => {
    //     resolve(200)
    //   })
    // }, err => {
    //   console.log('err: ', err);
    //   return '失败'
    //   return new Promise((resolve, reject) => {
    //     reject(404)
    //     // resolve(404)
    //   })
    // }).then(res => console.log('===', res), error => console.log(error));

    // 链式调用2个核心要点
    // 1. 上一个 .then 要返回一个 promise 对象
    // 2. 下一个 .then 的参数要拿到上一个 .then 回调的返回值


    // 循环引用
    // const p2 = p.then(res => {
    //   console.log('res: ', res);
    //   setTimeout(() => {
    //     return p2
    //   }, 0);
    // }, error => {
    //   console.log(error)
    //   return p2
    // });

    // Promise.resolve(4).then(res => console.log('===', res))
    // Promise.reject('静态方法reject').then(res => console.log('===', res), error => console.log('error', error))

    const p3 = new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve(3000)
      }, 3000);
    })

    const p4 = new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve(1000)
      }, 1000);
    })

    const p5 = new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve(2000)
      }, 2000);
    })

    const p6 = new Promise((resolve, reject) => {
      resolve(666)
    })

    Promise.all([p3, p4, p5, p6]).then(res => console.log('===', res))
    Promise.race([p3, p4, p5, p6]).then(res => console.log('===', res))
    Promise.race([Promise.reject('第一个被返回'), p3, p4, p5, p6]).then(res => console.log('===', res), error => console.log('error', error))
    Promise.race([Promise.resolve('第一个被返回'), p3, p4, p5, p6]).then()

  </script>
</body>

</html>